### PR TITLE
[IMP] pos_restaurant: floor background picture

### PR DESCRIPTION
**Pull request by loukwets**
**Before this PR:**
It was not possible to use a custom image for the background of a floor plan.

**Now:**
It is now possible to add an image to the floor plan in order to use it as the background. 
If a colour is selected when a custom image is used, the image will be removed from the record.

The custom background image cannot be used in kanban view of the floor plan.

image

Task-3616978

### CODE REVIEW

File: addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js [hunk 1]
```diff
@@ -616,6 +635,30 @@ export class FloorScreen extends Component {
...
+                });
+            } else {
+                this.dialog.add(AlertDialog, {
+                    title: _t("Loading Image Error"),
+                    body: _t("Encountered error when loading image. Please try again."),
+                });
+            }
+            this.state.isColorPicker = false;
+        }
+    }
```
Comment (davidmonnom) [inline addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js:661]: It seems that the server process the image in a different way, reload the image from the server after upload it.

```suggestion
    async uploadImage(event) {
        const file = event.target.files[0];
        if (!file.type.match(/image.*/)) {
            this.dialog.add(AlertDialog, {
                title: _t("Unsupported File Format"),
                body: _t("Only web-compatible Image formats such as .png or .jpeg are supported."),
            });
        } else {
            const imageUrl = await getDataURLFromFile(file);
            const loadedImage = await _loadImage(imageUrl);
            if (loadedImage) {
                await this.pos.data.ormWrite("restaurant.floor", [this.activeFloor.id], {
                    floor_background_image: imageUrl.split(",")[1],
                });
                await this.pos.data.read("restaurant.floor", [this.activeFloor.id]);
            } else {
                this.dialog.add(AlertDialog, {
                    title: _t("Loading Image Error"),
                    body: _t("Encountered error when loading image. Please try again."),
                });
            }
            this.state.isColorPicker = false;
        }
    }
```

File: addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js [hunk 2]
```diff
@@ -67,6 +69,16 @@ export class FloorScreen extends Component {
...
+    get floorBackroundURL() {
+        return this.activeFloor.floor_background_image
+            ? "data:image/png;base64," + this.activeFloor.floor_background_image
+            : false;
+    }
+    get floorBackround() {
+        return this.activeFloor.floor_background_image
+            ? "url(" + this.floorBackroundURL + ")"
+            : "none";
+    }
```
Comment (davidmonnom) [inline addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js:81]: Why there is two methods ? 
```suggestion
    get floorBackround() {
        return this.activeFloor.floor_background_image
            ? "data:image/png;base64," + this.activeFloor.floor_background_image
            : "none";
    }
```

File: addons/pos_restaurant/static/src/app/floor_screen/floor_screen.xml [hunk 3]
```diff
@@ -54,7 +63,7 @@
             <div t-on-click="onClickFloorMap" t-on-touchstart="_onPinchStart" t-on-touchmove="_onPinchMove" t-on-touchend="_onPinchEnd"
                 class="floor-map position-relative flex-grow-1 flex-shrink-1 flex-basis-0 w-auto h-100 overflow-auto"
                 t-ref="floor-map-ref"
-                t-attf-style="background: {{activeFloor ? activeFloor.background_color : 'none'}};">
+                t-attf-style="background: {{activeFloor ? activeFloor.background_color : 'none'}}; background-image: {{pos.floorPlanStyle == 'kanban' ? '' : floorBackround}}; background-size: auto; background-repeat: no-repeat;">
```
Comment (davidmonnom) [inline addons/pos_restaurant/static/src/app/floor_screen/floor_screen.xml:66]:
```suggestion
                t-attf-style="{{ activeFloor.floor_background_image ? 'background-image: url(' + floorBackround + '); background-size: auto; background-repeat: no-repeat;' : 'background: ' + activeFloor.background_color }}">
```

File: addons/pos_restaurant/static/src/app/floor_screen/floor_screen.xml [hunk 4]
```diff
@@ -39,6 +39,15 @@
                 <t t-set="lighter" t-value="selectedTables.length === 0" />
                 <button t-foreach="Object.entries(getColors())" t-as="color" t-key="color[0]" class="p-4 border-0"
                     t-attf-style="background-color: {{lighter ? getLighterShade(color[0]) : color[1]}}" t-on-click.stop="() => this.setColor(lighter ? this.getLighterShade(color[0]) : color[1])" />
+                <div class="floor-picture position-relative text-center overflow-hidden d-flex align-items-center justify-content-center">
+                    <t t-if="floorBackroundURL">
```
Comment (davidmonnom) [inline addons/pos_restaurant/static/src/app/floor_screen/floor_screen.xml:43]:
```suggestion
                    <t t-if="activeFloor.floor_background_image">
```

**Review by davidmonnom — COMMENTED** [review]
Ask me if you have questions

File: addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js [hunk 5]
```diff
@@ -616,6 +630,30 @@ export class FloorScreen extends Component {
...
+                body: _t("Only web-compatible Image formats such as .png or .jpeg are supported."),
+            });
+        } else {
+            const imageUrl = await getDataURLFromFile(file);
+            const loadedImage = await loadImage(imageUrl);
+            if (loadedImage) {
+                await this.pos.data.ormWrite("restaurant.floor", [this.activeFloor.id], {
+                    floor_background_image: imageUrl.split(",")[1],
+                });
+                await this.pos.data.read("restaurant.floor", [this.activeFloor.id]);
```
Comment (davidmonnom) [inline addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js:647]: Maybe just put a comment about why write + read are needed.

**Comment by davidmonnom** [comment]
LGTM, but what do you think about adding a loader indicator during image upload ? If you test on the runbot it can take 10 secondes.

File: addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js [hunk 6]
```diff
@@ -28,19 +31,85 @@ export class FloorScreen extends Component {
...
+            ...tables.map((table) => table.position_h + table.width),
+            floorH
+        );
+        const positionV = Math.max(
+            ...tables.map((table) => table.position_v + table.height),
+            floorV
+        );
+
+        if (this.activeFloor.floor_background_image) {
+            const i = new Image();
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js:98]:
```suggestion
            const img = new Image();
```
use a more sensible variable name

File: addons/point_of_sale/static/src/app/screens/product_screen/product_screen.js [hunk 7]
```diff
@@ -95,29 +95,21 @@ export class ProductScreen extends Component {
...
+            isSelected: this.getAncestorsAndCurrent().includes(category) && !this.ui.isSmall,
             isChildren: this.getChildCategories(this.pos.selectedCategory).includes(category),
         }));
     }
-
+    getCategoriesAndSub() {
+        return this.getAncestorsAndCurrent().flatMap((category) =>
+            this.getChildCategoriesInfo(category)
+        );
+    }
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/screens/product_screen/product_screen.js:112]: restore the original location of this code to reduce diff.

File: addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js [hunk 8]
```diff
@@ -28,19 +31,85 @@ export class FloorScreen extends Component {
...
+                        this.floorScrollBox.el.classList.remove("overflow-scroll");
+                        this.floorScrollBox.el.classList.add("overflow-hidden");
+                    } else {
+                        this.floorScrollBox.el.classList.remove("overflow-hidden");
+                        this.floorScrollBox.el.classList.add("overflow-scroll");
+                    }
+                }
+            },
+            () => [this.state.selectedTableIds]
+        );
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js:77]:
```suggestion
        useEffect(
            (nSelectedTables) => {
                if (hasTouch()) {
                    if (nSelectedTables) {
                        this.floorScrollBox.el.classList.remove("overflow-scroll");
                        this.floorScrollBox.el.classList.add("overflow-hidden");
                    } else {
                        this.floorScrollBox.el.classList.remove("overflow-hidden");
                        this.floorScrollBox.el.classList.add("overflow-scroll");
                    }
                }
            },
            () => [this.state.selectedTableIds.length]
        );
```

File: addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js [hunk 9]
```diff
@@ -28,19 +31,85 @@ export class FloorScreen extends Component {
...
+        useEffect(
+            () => {
+                document.addEventListener("resize", this.computeFloorSize);
+                return () => {
+                    document.removeEventListener("resize", this.computeFloorSize);
+                };
+            },
+            () => []
+        );
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js:63]:
```suggestion
        useEffect(
            () => {
                const computeFloorSize = () => this.computeFloorSize();
                document.addEventListener("resize", computeFloorSize);
                return () => {
                    document.removeEventListener("resize", computeFloorSize);
                };
            },
            () => []
        );
```
make sure the callback is properly bound to `this`.

**Pull request closed without merging**